src/MongoDB.Driver.Core/Core/Operations/AsyncCursor.cs

```diff
@@ -247,24 +327,79 @@ public void Dispose()
         /// </summary>
         /// <param name="disposing"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
         protected virtual void Dispose(bool disposing)
+        {
+            try
+            {
+                DisposeCursor(disposing);
+            }
+            catch 
+            {
+                // ignore exceptions
+            }
+        }
+
+        private void DisposeCursor(bool disposing, CancellationToken cancellationToken = default(CancellationToken))
```

Comment:
Let's call this method CloseCursor and deal ONLY with closing the cursor (let Dispose handle Dispose related tasks):

    private void CloseCursor(CancellationToken cancellationToken)
    {
        if (!_closed)
        {
            try
            {
                if (_cursorId != 0)
                {
                    KillCursors(cancellationToken);
                }
            }
            finally
            {
                _closed = true;
            }
        }
    }

Note that this method is designed so that if called more than once subsequent calls are a no-op.